google/cloud/aiplatform/base.py

```diff
@@ -561,6 +578,141 @@ def _sync_object_with_future_result(
+
+        self = cls._empty_constructor(
+            project=project, location=location, credentials=credentials
+        )
+
+        # Fetch credentials once and re-use for all `_empty_constructor()` calls
+        creds = initializer.global_config.credentials
+
+        resource_list_method = getattr(self.api_client, self._list_method)
+
+        list_request = {
+            "parent": initializer.global_config.common_location_path(),
+            "filter": filter,
+        }
+
+        if order_by:
+            list_request["order_by"] = order_by
+
+        resource_list = resource_list_method(request=list_request) or []
+
+        return [
+            self._construct_sdk_resource_from_gapic(
+                gapic_resource, project=project, location=location, credentials=creds
+            )
+            for gapic_resource in resource_list
+            if cls_filter(gapic_resource)
+        ]
+
+    @classmethod
+    def _list_with_local_order(
+        cls,
+        cls_filter: Callable[[proto.Message], bool] = lambda _: True,
+        filter: Optional[str] = None,
+        order_by: Optional[str] = None,
+        project: Optional[str] = None,
+        location: Optional[str] = None,
+        credentials: Optional[auth_credentials.Credentials] = None,
+    ) -> List[AiPlatformResourceNoun]:
+        """Client-side sorting when list API doesn't support `order_by`"""
+
+        li = cls._list(
+            cls_filter=cls_filter,
+            filter=filter,
+            order_by=None,  # This method will handle the ordering locally
+            project=project,
+            location=location,
+            credentials=credentials,
+        )
+
+        desc = "desc" in order_by
+        order_by = order_by.replace("desc", "")
+        order_by = order_by.split(",")
+
+        li.sort(
+            key=lambda x: tuple(getattr(x, field.strip()) for field in order_by),
+            reverse=desc,
+        )
+
+        return li
+
+    @classmethod
+    def list(
+        cls,
+        filter: Optional[str] = None,
+        order_by: Optional[str] = None,
+        project: Optional[str] = None,
+        location: Optional[str] = None,
+        credentials: Optional[auth_credentials.Credentials] = None,
+    ) -> List[AiPlatformResourceNoun]:
+        """List all instances of this AI Platform Resource.
+
+        Example Usage:
+
+        aiplatform.BatchPredictionJobs.list(
+            filter='state="JOB_STATE_SUCCEEDED" AND display_name="my_job"',
+        )
+
+        aiplatform.Model.list(order_by="create_time desc, display_name")
+
+        Args:
+            filter (str):
+                Optional. An expression for filtering the results of the request.
+                For field names both snake_case and camelCase are supported.
+            order_by (str):
+                Optional. A comma-separated list of fields to order by, sorted in
+                ascending order. Use "desc" after a field name for descending.
```

Comment:
Is something like `'display_name, desc, create_time, update_time, desc'` valid here or it only when `desc` is at the end of the list? It doesn't seem like the local logic handles that case.

Reply:
`desc` is only considered valid by the service when it follows one of the three fields and remains inside the comma delim. Manually tried other combinations, I think the local logic handles all cases that are valid service-side.

```python
In [5]: endpoint_list = aiplatform.Model.list(order_by='display_name, desc, create_time, update_time, desc')
   ...: print_list(endpoint_list)
---------------------------------------------------------------------------
... long stack trace here ...

~/mbsdk/vinny-aip/venv/lib/python3.7/site-packages/google/api_core/grpc_helpers.py in error_remapped_callable(*args, **kwargs)
     57             return callable_(*args, **kwargs)
     58         except grpc.RpcError as exc:
---> 59             six.raise_from(exceptions.from_grpc_error(exc), exc)
     60 
     61     return error_remapped_callable

~/mbsdk/vinny-aip/venv/lib/python3.7/site-packages/six.py in raise_from(value, from_value)

InvalidArgument: 400 Provided `order_by` field has invalid format.
```